File: 20/aoc7.cc
#include <algorithm>
#include <iostream>
#include <fstream>
#include <map>
#include <string>
#include <vector>
#include "aoc_helper.h"

// part 1
// std::map<std::string, std::vector<std::string>> parseBags(std::string row) {
//     std::map<std::string, std::vector<std::string>> entry;
//     // vibrant aqua bags contain 1 shiny magenta bag, 2 muted teal bags, 1 dim magenta bag, 1 muted chartreuse bag.
//     std::vector<std::string> splitByContain = splitString(row, "bags contain");
//     if (splitByContain.size() != 2) {
//         std::cout << "There was a problem with your input data: " << row << std::endl;
//         exit(1);
//     }
//     std::string parent = strip(splitByContain[0]);
//     // std::cout << "container: " << key << " size: " << key.size() << std::endl;
//     entry[parent] = {};
//     //parse the containing bags
//     std::vector<std::string> splitByCommas = splitString(splitByContain[1], ",");
//     for (int i = 0; i < splitByCommas.size(); i++) {
//         std::string unparsed = splitByCommas[i];
//         std::string noBags = removeFromString(unparsed, {"bags", "bag", "."});
//         std::string stripped = strip(noBags);
//         // std::cout << stripped << " size: " << stripped.size() << std::endl;
//         if (noBags.find("no other") != std::string::npos) {
//             continue;
//         }
//         if ( entry.find(stripped) == entry.end() ) {
//             // not found
//                 entry[stripped] = {parent};
//         } else {
//         // found
//             entry[stripped].push_back(parent);
//         }
//     }
//     return entry;
// }

// part 2
std::map<std::string, std::vector<std::pair<std::string, int>>> parseBags(std::string row) {
    std::map<std::string, std::vector<std::pair<std::string, int>>> entry;
    // vibrant aqua bags contain 1 shiny magenta bag, 2 muted teal bags, 1 dim magenta bag, 1 muted chartreuse bag.
    std::vector<std::string> splitByContain = splitString(row, "bags contain");
    if (splitByContain.size() != 2) {
        std::cout << "There was a problem with your input data: " << row << std::endl;
        exit(1);
    }
    std::string parent = strip(splitByContain[0]);
    // std::cout << "container: " << key << " size: " << key.size() << std::endl;
    entry[parent] = {};
    //parse the containing bags
    std::vector<std::string> splitByCommas = splitString(splitByContain[1], ",");
    for (int i = 0; i < splitByCommas.size(); i++) {
        std::string unparsed = splitByCommas[i];
        std::string noBags = removeFromString(unparsed, {"bags", "bag", "."});
        std::string stripped = strip(noBags);
        // https://stackoverflow.com/a/30727561/4458404
        int n = stripped[0] - '0'; // convert to int
        stripped = strip(stripped.substr(1, stripped.size()- 1));
        
        if (noBags.find("no other") != std::string::npos) {
            continue;
        }
        if ( entry.find(parent) == entry.end() ) {
            // not found
                entry[parent] = {std::pair<std::string,int> (stripped, n)};
        } else {
        // found
            entry[parent].push_back(std::pair<std::string,int> (stripped, n));
        }

    }
    return entry;
}

int childrenBags( std::map<std::string, std::vector<std::pair<std::string, int>>> adjacency_list, std::string target, int bagN) {
    int count = 0;
    std::vector<std::pair<std::string, int>> children = adjacency_list[target];
    std::cout << target << " count is " << bagN << std::endl;
    count += bagN;
    for (int i = 0; i < children.size(); i++) {
        std::string child = children[i].first;
        std::cout << "There are " << children[i].second << " " << child << " bags in a " << target << std::endl;
        int nestedBags = childrenBags(adjacency_list, child, children[i].second);
        std::cout << target << " contains " <<  nestedBags << std::endl;
        count +=  bagN * nestedBags;
    }
    return count;
}

int baggageClaim(std::vector<std::string> arr, std::string target) {
    std::string row;
    std::map<std::string, std::vector<std::pair<std::string, int>>> adjacency_list;
    for(int i = 0; i < arr.size(); i++) {
        row = arr[i];
        std::map<std::string, std::vector<std::pair<std::string, int>>> entry = parseBags(row);
        std::map<std::string, std::vector<std::pair<std::string, int>>>::iterator it;
        for ( it = entry.begin(); it != entry.end(); it++ )
        {
            adjacency_list[it->first] = it->second;
        }
    }    
    std::map<std::string, bool> parent_check;
    int bags = childrenBags(adjacency_list, target, 1);
    return bags;
}


int main() {
    //read input from file
    std::vector<std::string> data;
    data = getInputString("./input/aoc7.txt");

    int parentBags = baggageClaim(data, "shiny gold");
    //need to subtract one to not count the target
    std::cout << "the number of parent bags " << parentBags - 1 << std::endl;
}

File: 20/python/aoc7.py
def count_parents(lug_map, target):
    to_search = [target]
    prospective_parents = {}
    while len(to_search) > 0:
        current_target = to_search.pop()
        if current_target in lug_map:
            targets = lug_map[current_target]
        else:
            continue

        for new_target in targets:
            prospective_parents[new_target] = True
            to_search.append(new_target)
    return len(prospective_parents)

def luggage(data):
    luggage_map = {}
    for line in data:
        line = line.strip()
        split_by_contain = line.split("bags contain")
        parent_bag = split_by_contain[0].strip()
        if "no other bags" in split_by_contain[1]:
            luggage_map[parent_bag] = []
            continue
        split_by_comma = split_by_contain[1].split(",")
        for bag in split_by_comma:
            bag = bag.strip()
            bag_split = bag.split(" ")
            child_bag = bag_split[1] + " " + bag_split[2]
            if child_bag in luggage_map:
                luggage_map[child_bag].append(parent_bag)
            else:
                luggage_map[child_bag] = [parent_bag]
        
    return count_parents(luggage_map, "shiny gold")



def count_children(lug_map, target):
    # assume target in map
    if len(lug_map[target]) == 0:
        return 0
    count = 0
    for curr_target in lug_map[target]:
        count += int(curr_target[1])
        count += (int(curr_target[1]) * count_children(lug_map, curr_target[0]))
    return count

def luggage2(data):
    luggage_map = {}
    for line in data:
        line = line.strip()
        split_by_contain = line.split("bags contain")
        parent_bag = split_by_contain[0].strip()
        if "no other bags" in split_by_contain[1]:
            luggage_map[parent_bag] = []
            continue
        split_by_comma = split_by_contain[1].split(",")
        for bag in split_by_comma:
            bag = bag.strip()
            bag_split = bag.split(" ")
            child_bag = bag_split[1] + " " + bag_split[2]
            child_bag_count = bag_split[0]
            if parent_bag in luggage_map:
                luggage_map[parent_bag].append((child_bag, child_bag_count))
            else:
                luggage_map[parent_bag] = [(child_bag, child_bag_count)]
        
    return count_children(luggage_map, "shiny gold")

if __name__ == "__main__":
    text_file = open("../input/aoc7.txt", "r")
    string_data = text_file.readlines()
    ans = luggage2(string_data)
    print("the number of bags a shiny gold bags contains is {}".format(ans))
